File: services/shop_service.py
import time
from datetime import datetime, timedelta
from database import db
from utils.data.items import ITEMS

class ShopService:
    async def refresh_shop_if_needed(self, user):
        now_ms = int(time.time() * 1000)
        
        # Start of today (UTC)
        now_date = datetime.utcnow()
        start_of_day = datetime(now_date.year, now_date.month, now_date.day)
        start_of_day_ms = int(start_of_day.timestamp() * 1000)
        
        # Start of week (Monday)
        days_to_monday = now_date.weekday() # 0 is Monday
        start_of_week = start_of_day - timedelta(days=days_to_monday)
        start_of_week_ms = int(start_of_week.timestamp() * 1000)

        shop_state = user.get('shopState', {
            'daily': {'last': 0, 'stock': {}},
            'weekly': {'last': 0, 'stock': {}}
        })

        updated = False

        # Daily Refresh
        if shop_state.get('daily', {}).get('last', 0) < start_of_day_ms:
            shop_state['daily'] = {'last': now_ms, 'stock': {}}
            for item_id, item in ITEMS.items():
                if item.get('shop', {}).get('category') == 'daily':
                    shop_state['daily']['stock'][item_id] = item['shop']['stock']
            updated = True

        # Weekly Refresh
        if shop_state.get('weekly', {}).get('last', 0) < start_of_week_ms:
            shop_state['weekly'] = {'last': now_ms, 'stock': {}}
            for item_id, item in ITEMS.items():
                if item.get('shop', {}).get('category') == 'weekly':
                    shop_state['weekly']['stock'][item_id] = item['shop']['stock']
            updated = True

        if updated:
            user['shopState'] = shop_state
            await db.users.update({"telegramId": user['telegramId']}, {"$set": {"shopState": shop_state}})
        
        return shop_state

    async def buy_item(self, user_id, item_id):
        user = await db.users.find_one({"telegramId": user_id})
        item = ITEMS.get(item_id)

        if not user or not item:
            return {"success": False, "msg": "Transaction Error: Data Mismatch."}
        
        await self.refresh_shop_if_needed(user)
        category = item.get('shop', {}).get('category', 'special')
        
        # 1. Stock Check
        if category != 'special':
            current_stock = user['shopState'].get(category, {}).get('stock', {}).get(item_id, 0)
            if current_stock <= 0:
                return {"success": False, "msg": "❌ This item is sold out! Check back after refresh."}

        # 2. Currency Check
        total_cost = item['price']
        currency = item.get('currency', 'coins')
        if user.get(currency, 0) < total_cost:
            return {"success": False, "msg": f"❌ Not enough {currency}! You need {total_cost - user.get(currency, 0)} more."}

        # 3. Deduction
        user[currency] -= total_cost
        if category != 'special':
            user['shopState'][category]['stock'][item_id] -= 1

        # 4. Inventory Addition
        inventory = user.get('inventory', [])
        found = False
        for inv_item in inventory:
            if inv_item['id'] == item_id:
                inv_item['qty'] += 1
                found = True
                break
        
        if not found:
            inventory.append({"id": item_id, "qty": 1})
        
        user['inventory'] = inventory

        # Special Field Updates
        if item_id == 'gacha_ticket':
            user['gachaTickets'] = user.get('gachaTickets', 0) + 1
        elif item_id == 'gacha_pack':
            user['gachaTickets'] = user.get('gachaTickets', 0) + 10

        # Save to DB
        await db.users.update({"telegramId": user_id}, {"$set": {
            currency: user[currency],
            "shopState": user['shopState'],
            "inventory": user['inventory'],
            "gachaTickets": user.get('gachaTickets', 0)
        }})

        return {
            "success": True,
            "item": item,
            "qty": 1,
            "totalCost": total_cost,
            "currency": currency,
            "msg": f"✅ Successfully purchased 1x {item['name']}!"
        }

shop_service = ShopService()

File: services/shopService.js
const db = require('../database');
const items = require('../utils/data/items');

/**
 * Shop Service: Handles secure transactions, stock management, and timed refreshes.
 */
module.exports = {
    /**
     * Checks and refreshes shop stock if needed based on timeframe.
     */
    async refreshShopIfNeeded(user) {
        const now = Date.now();
        const startOfDay = new Date().setUTCHours(0, 0, 0, 0);
        
        // Find last Monday at 00:00 UTC
        const today = new Date();
        const day = today.getUTCDay(); // 0 is Sunday, 1 is Monday
        const diff = (day === 0 ? 6 : day - 1);
        const startOfWeek = new Date(today.setUTCDate(today.getUTCDate() - diff)).setUTCHours(0, 0, 0, 0);

        user.shopState = user.shopState || { daily: { last: 0, stock: {} }, weekly: { last: 0, stock: {} } };

        let updated = false;

        // Daily Refresh
        if (user.shopState.daily.last < startOfDay) {
            user.shopState.daily.last = now;
            user.shopState.daily.stock = {};
            // Initialize stock from metadata
            Object.values(items).filter(i => i.shop && i.shop.category === 'daily').forEach(i => {
                user.shopState.daily.stock[i.id] = i.shop.stock;
            });
            updated = true;
        }

        // Weekly Refresh
        if (user.shopState.weekly.last < startOfWeek) {
            user.shopState.weekly.last = now;
            user.shopState.weekly.stock = {};
            Object.values(items).filter(i => i.shop && i.shop.category === 'weekly').forEach(i => {
                user.shopState.weekly.stock[i.id] = i.shop.stock;
            });
            updated = true;
        }

        if (updated) {
            await db.users.update({ telegramId: user.telegramId }, { $set: { shopState: user.shopState } });
        }
        return user.shopState;
    },

    /**
     * Attempts to purchase an item with stock and balance checks.
     */
    async buyItem(userId, itemId) {
        let user = await db.users.findOne({ telegramId: userId });
        const item = items[itemId];

        if (!user || !item) return { success: false, msg: "Transaction Error: Data Mismatch." };
        
        await this.refreshShopIfNeeded(user);
        const category = item.shop.category;
        
        // 1. Stock Check
        if (category !== 'special') {
            const currentStock = user.shopState[category].stock[itemId] || 0;
            if (currentStock <= 0) return { success: false, msg: "❌ This item is sold out! Check back after refresh." };
        }

        // 2. Currency Check
        const totalCost = item.price;
        const currency = item.currency; // 'coins'
        if ((user[currency] || 0) < totalCost) {
            return { success: false, msg: `❌ Not enough ${currency}! You need ${totalCost - (user[currency] || 0)} more.` };
        }

        // 3. Deduction
        user[currency] -= totalCost;
        if (category !== 'special') {
            user.shopState[category].stock[itemId] -= 1;
        }

        // 4. Inventory Addition
        user.inventory = user.inventory || [];
        const invIndex = user.inventory.findIndex(i => i.id === itemId);
        if (invIndex > -1) {
            user.inventory[invIndex].qty += 1;
        } else {
            user.inventory.push({ id: itemId, qty: 1 });
        }

        // Special: Gacha Tickets (Direct field update)
        if (itemId === 'gacha_ticket') {
            user.gachaTickets = (user.gachaTickets || 0) + 1;
        } else if (itemId === 'gacha_pack') {
            user.gachaTickets = (user.gachaTickets || 0) + 10;
        }

        await db.users.update({ telegramId: userId }, { $set: user });

        return { 
            success: true, 
            item, 
            qty: 1, 
            totalCost, 
            currency,
            msg: `✅ Successfully purchased 1x ${item.name}!`
        };
    }
};
